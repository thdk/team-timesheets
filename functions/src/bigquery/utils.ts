// Imports the Google Cloud client library
import { BigQuery } from '@google-cloud/bigquery';
import { BigQueryTableSchemes } from './config';

export type BigQueryField = { name: string, type: string, mode?: string };
export type BigQueryConfig = {
    dataSetId: string;
    tableId: string;
    tableIdPrefix: string,
    schemes: BigQueryTableSchemes,
    timePartitioningField?: string,
};

function validateTableSchemaAsync(table: any, schema: BigQueryField[]) {
    return table.getMetadata().then(([metadata]) => {
        if (metadata.schema) {
            console.log(`Current schema of table ${table.id} is ${JSON.stringify(metadata.schema.fields)}`);
        } else {
            console.log("Table has no schema yet.");
        }

        console.log(`Required schema of table ${table.id} is ${JSON.stringify(schema)}`);

        if (!isEqualSchema(metadata.schema.fields, schema)) {
            console.log(`Updating schema of table ${table.id}...`);
            return table.setMetadata(Object.assign(metadata, { schema }))
                .then(() => {
                    console.log("Schema updated.");
                    return table;
                }, () => {
                    console.log("Couldn't update schema of table: " + table.id);
                });
        }
        else {
            console.log(`Schema of table ${table.id}... is already up to date.`);
            return new Promise(resolve => resolve(table));
        }
    });
}

function isEqualSchema(schema1: BigQueryField[], schema2: BigQueryField[]) {
    console.log(`Comparing fields length... ${schema1.length} <> ${schema2.length}`);
    if (schema1.length !== schema2.length) return false;

    return !schema1.some((field1, index) => {

        const field2 = schema2[index];
        console.log(`Comparing field...`);
        console.log(`${JSON.stringify(field1)}`);
        console.log(`${JSON.stringify(field2)}`);

        return field1.name !== field2.name ||
            field1.type !== field2.type ||
            field1.mode !== field2.mode;
    });
}

type InsertOptions = {
    schema: BigQueryField[];
    location: string;
    timePartitioning?: {
        type: 'DAY';
        field: string;
    };
}
export function insertRowsAsync<T>(options: BigQueryConfig, rows: ReadonlyArray<T>, bigqueryClient?: BigQuery): Promise<void> {
    const { dataSetId, tableId, tableIdPrefix = "", schemes, timePartitioningField } = options;

    console.log(`Inserting ${rows.length} rows into ${tableId}`);

    let insertOptions: InsertOptions = {
        schema: schemes[tableId],
        location: 'US'
    };

    if (timePartitioningField) {
        insertOptions = {
            ...insertOptions,
            timePartitioning: {
                type: 'DAY',
                field: timePartitioningField,
            },
        }
    }

    if (!rows.length) return new Promise(resolve => resolve());

    const bigquery = bigqueryClient || new BigQuery();

    // TODO: create dataset only once since insertRows can be called multiple times simultaneously!
    return bigquery
        .dataset(dataSetId)
        .get({ autoCreate: true })
        .then(([dataset]) => {
            console.log("dataset created");
            return dataset.table(`${tableIdPrefix}${tableId}`)
                .get({ autoCreate: true, ...insertOptions })
                .then(([table]) => validateTableSchemaAsync(table, insertOptions.schema))
                .then(table => {
                    const first10kRows = rows.slice(0, 10000);
                    return table.insert(first10kRows)
                        .then(() => {
                            console.log(`Inserted ${first10kRows.length} rows of total ${rows.length} into ${tableId}`);
                            if (rows.length > 10000) {
                                console.log("Inserting next batch...")
                                return insertRowsAsync(options, rows.slice(10000));
                            }

                            console.log(`Finished inserting into ${tableId}`);
                            return true;
                        }, (error: { errors: any, name: string, response: any, message: string }) => {
                            error.errors && error.errors.forEach(e => {
                                console.log({
                                    error: error.name,
                                    message: error.message,
                                    errors: e.errors,
                                    row: e.row
                                });
                            });
                        }, (e) => {
                            console.log(e);
                            console.log("Table could not be created");
                        })
                })
        }, e => {
            console.log(e);
            console.log("Dataset could not be created");
        });

}