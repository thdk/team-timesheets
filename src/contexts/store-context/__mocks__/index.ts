export const useStore = jest.fn();
