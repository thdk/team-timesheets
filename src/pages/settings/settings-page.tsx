import React, { useCallback } from 'react';

import { TaskList } from '../../containers/tasks/list';
import { SettingsTab, SettingsRouteQueryParams } from '../../routes/settings';
import { goToSettings } from '../../internal';
import { ClientList } from '../../containers/clients/list';
import { UserList } from '../../containers/users/list';
import { canReadUsers, canManageTeams } from '../../rules';
import { TeamList } from '../../containers/teams/list';
import { useRouterStore } from '../../stores/router-store';
import { observer } from 'mobx-react-lite';
import { useUserStore } from "../../contexts/user-context";
import { Connections } from './connections';
import { Tabs, ITabData } from '../../components/tabs';

export const SettingsPage = observer(() => {
    const router = useRouterStore();

    const { divisionUser: user } = useUserStore();

          const tabData: ITabData<SettingsTab>[] = [
              { id: "tasks", text: "Tasks", tabContent: <TaskList /> },
              { id: "clients", text: "Clients", tabContent: <ClientList /> },
              { id: "teams", text: "Teams", canOpen: () => canManageTeams(user), tabContent: <TeamList /> },
              { id: "users", text: "Users", canOpen: () => canReadUsers(user), tabContent: <UserList /> },
              { id: "connections", text: "Connections", tabContent: <Connections /> },
        ];

    const onTabChange = useCallback((tabId: SettingsTab) => {
        goToSettings(router, tabId);
    }, [goToSettings, router]);

    const query: { tab: SettingsTab } = router.queryParams as SettingsRouteQueryParams;
    const { tab: activeTabId = undefined } = query || {};

    return (
        <Tabs
            tabData={tabData}
            activeTab={activeTabId}
            onTabChange={onTabChange}
        />
    );
});
