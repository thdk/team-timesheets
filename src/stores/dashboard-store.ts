import { observable, reaction, computed, action, makeObservable } from "mobx";
import { ICollection, Collection, RealtimeMode } from "firestorable";

import moment from 'moment';

import { IRootStore } from "./root-store";
import { IGroupedRegistrations } from "./registration-store/registration-store";
import * as deserializer from '../../common/serialization/deserializer';
import * as serializer from '../../common/serialization/serializer';
import { TimePeriod } from "../components/time-period-select";
import { IRegistration, IRegistrationData } from "../../common/dist";
import { Firestore, query, Query, where } from "firebase/firestore";

export interface IDashboardStore {
    readonly setProjectFilter: (projectId: string | undefined) => void;
    readonly setUserFilter: (userId: string | undefined) => void;
    readonly setTaskFilter: (taskId: string | undefined) => void;
    readonly setTimePeriodFilter: (timePeriod: TimePeriod) => void;

    readonly registrationsGroupedByUser: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByTask: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByProject: IGroupedRegistrations<string>[];

    readonly projectFilterValue?: string;
    readonly taskFilterValue?: string;
    readonly userFilterValue?: string;
    readonly timePeriodFilterValue?: TimePeriod;

}

export class DashboardStore implements IDashboardStore {
    private readonly registrationsField: ICollection<IRegistration, IRegistrationData>;

    private userFilterValueField: string | undefined = undefined;
    private taskFilterValueField: string | undefined = undefined;
    private projectFilterValueField: string | undefined = undefined;
    private timePeriodFilterField: TimePeriod | undefined = undefined;

    constructor(
        rootStore: IRootStore,
        {
            firestore,
        }: {
            firestore: Firestore,
        },
    ) {
        makeObservable<DashboardStore, "userFilterValueField" | "taskFilterValueField" | "projectFilterValueField" | "timePeriodFilterField" | "startDateFilter" | "endDateFilter" | "userIdFilter" | "projectIdFilter" | "startDate" | "endDate" | "registrations">(this, {
            userFilterValueField: observable.ref,
            taskFilterValueField: observable.ref,
            projectFilterValueField: observable.ref,
            timePeriodFilterField: observable.ref,
            startDateFilter: computed,
            endDateFilter: computed,
            userIdFilter: computed,
            projectIdFilter: computed,
            registrationsGroupedByUser: computed,
            registrationsGroupedByProject: computed,
            registrationsGroupedByTask: computed,
            startDate: computed,
            endDate: computed,
            registrations: computed,
            setProjectFilter: action,
            setTaskFilter: action,
            setUserFilter: action,
            setTimePeriodFilter: action
        });

        this.registrationsField = new Collection<IRegistration, IRegistrationData>(
            firestore,
            "registrations",
            {
                realtimeMode: RealtimeMode.on,
                deserialize: deserializer.convertRegistration,
                serialize: serializer.convertRegistration,
                name: "Dashboard registrations",
            },
            {
                // logger: console.log
            },
        );

        const updateRegistrationQuery = () => {
            if (!rootStore.user.divisionUser || !this.timePeriodFilterField) {
                this.registrationsField.query = null;
            } else {
                this.registrationsField.query = ref =>
                    this.userIdFilter(
                        this.projectIdFilter(
                            this.endDateFilter(
                                this.startDateFilter(
                                    query(ref, where("deleted", "==", false))
                                )
                            )
                        )
                    ) as Query<IRegistrationData>;
            }
        };

        // update the query of the registration collection each time...
        // -- a filter changes (userId, startDate, endDate)
        // TODO: add a Filter button in UI to group all filter changes into one new search query
        reaction(() => this.timePeriodFilterField, updateRegistrationQuery);
        reaction(() => this.userFilterValue, updateRegistrationQuery);
        reaction(() => this.projectFilterValue, updateRegistrationQuery);
        reaction(() => rootStore.user.divisionUser, updateRegistrationQuery);
    }

    private get startDateFilter() {
        const startDate = this.startDate;
        return (q: Query) => startDate ? query(q, where("date", ">=", startDate)) : q;
    }

    private get endDateFilter() {
        const endDate = this.endDate;
        return (q: Query) => endDate ? query(q, where("date", "<=", endDate)) : q;
    }

    private get userIdFilter() {
        const userId = this.userFilterValue;
        return (q: Query) => userId ? query(q, where("userId", "==", userId)) : q;
    }

    private get projectIdFilter() {
        return (q: Query) => this.projectFilterValue
            ? query(q, where("project", "==", this.projectFilterValue))
            : q;
    }

    get registrationsGroupedByUser() {
        // no need to group when userId is included in the filter
        if (this.registrations.length === 0) return [];

        const registrationsPerUser = this.createRegistrationsGroup(reg => reg.userId);

        return registrationsPerUser;
    }

    get registrationsGroupedByProject() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.project || "")
            .sort((a, b) => b.totalTime - a.totalTime);;
    }

    get registrationsGroupedByTask() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.task || "")
            .sort((a, b) => b.totalTime - a.totalTime);
    }

    private get startDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().startOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().startOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').startOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().startOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').startOf('year').toDate();
            default:
                return undefined;
        }
    }

    private get endDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().endOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().endOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').endOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().endOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').endOf('year').toDate();
            default:
                return undefined;
        }
    }

    public get taskFilterValue() {
        return this.taskFilterValueField;
    }

    public get userFilterValue() {
        return this.userFilterValueField;
    }

    public get projectFilterValue() {
        return this.projectFilterValueField;
    }

    public get timePeriodFilterValue() {
        return this.timePeriodFilterField;
    }

    private get registrations() {
        return this.registrationsField.docs
            .filter(doc => doc.data!.isPersisted) // don't display drafts (TODO: move to collection.ts?)
    }

    public setProjectFilter(projectId: string | undefined) {
        this.projectFilterValueField = projectId;
    }

    public setTaskFilter(taskId: string | undefined) {
        this.taskFilterValueField = taskId;
    }

    public setUserFilter(userId: string | undefined) {
        this.userFilterValueField = userId;
    }

    setTimePeriodFilter(timePeriod: TimePeriod) {
        this.timePeriodFilterField = timePeriod;
    }

    private createRegistrationsGroup = <T>(keySelector: (reg: IRegistration) => T): IGroupedRegistrations<T>[] => {
        const groupedMap = new Map<T, IGroupedRegistrations<T>>();

        this.registrations
            .forEach(c => {
                const key = keySelector(c.data!);
                const group = groupedMap.get(key);
                if (group) {
                    group.totalTime += c.data!.time || 0;
                    group.registrations.push(c);
                }
                else {
                    groupedMap.set(key, {
                        groupKey: keySelector(c.data!),
                        registrations: [c],
                        totalTime: c.data!.time || 0,
                        isCollapsed: true
                    });
                }
            });

        return Array.from(groupedMap.values());
    }
}