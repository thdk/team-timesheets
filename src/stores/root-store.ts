import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { IProjectStore, ProjectStore } from "./project-store";
import { FavoriteStore } from "./favorite-store";

export interface IRootStore {
    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly favorites: FavoriteStore;
    readonly dashboard: IDashboardStore;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;
    public readonly favorites: FavoriteStore;

    constructor() {
        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);
        this.favorites = new FavoriteStore(this);
    }
};
