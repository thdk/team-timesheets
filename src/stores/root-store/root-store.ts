import { RouterStore } from "mobx-router";

import { IRegistrationsStore, RegistrationStore } from "../registration-store/registration-store";
import { IConfigStore, ConfigStore } from "../config-store";
import { IUserStore, UserStore } from "../user-store";
import { IViewStore, ViewStore } from "../view-store";
import { IReportStore, ReportStore } from "../report-store/report-store";
import { DashboardStore, IDashboardStore } from "../dashboard-store";
import { ProjectStore } from "../project-store";
import { FavoriteStore } from "../favorite-store";
import { DivisionStore  } from "../division-store/division-store";

export interface IRootStore extends Store { };

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router: RouterStore<IRootStore>;
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: ProjectStore;
    public readonly favorites: FavoriteStore;
    public readonly divisions: DivisionStore;

    constructor({
        auth,
        firestore,
        httpsCallable,
    }: {
        firestore: firebase.firestore.Firestore,
        httpsCallable?: (name: string) => firebase.functions.HttpsCallable,
        auth?: firebase.auth.Auth,
    }) {
        this.user = new UserStore(
            this,
            {
                auth,
                firestore,
            }
        );

        this.view = new ViewStore();
        this.config = new ConfigStore(
            this,
            {
                firestore,
            },
        );
        this.timesheets = new RegistrationStore(
            this,
            {
                firestore,
            },
        );
        this.reports = new ReportStore(
            this,
            {
                firestore,
            },
        );
        this.dashboard = new DashboardStore(
            this,
            {
                firestore,
            },
        );
        this.projects = new ProjectStore(
            this,
            {
                firestore,
            },
        );
        this.favorites = new FavoriteStore(
            this,
            {
                firestore,
            },
        );
        this.divisions = new DivisionStore(
            this,
            {
                firestore,
                httpsCallable,
            },
        );
        this.router = new RouterStore<IRootStore>(this);
    }

    public dispose() {
        this.config.dispose();
        // this.dashboard.dispose();
        // this.favorites.dispose();
        // this.projects.dispose();
        // this.reports.dispose();
        this.timesheets.dispose();
        this.user.dispose();
        // this.view.dispose();
    }
};
