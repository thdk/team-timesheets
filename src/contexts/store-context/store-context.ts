import { createContext } from 'react';
import { IRootStore } from "../../stores/root-store";

export const StoreContext = createContext<IRootStore>({} as IRootStore);
export const StoreProvider = StoreContext.Provider;
