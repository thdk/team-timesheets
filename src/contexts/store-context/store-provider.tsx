import React from "react";
import { useFirebase } from "../firebase-context/firebase-context";
import { StoreContext } from "./store-context";
import { Store, IRootStore } from "../../stores/root-store";
import { getAuth } from "firebase/auth";
import { getFirestore } from "firebase/firestore";
import { getFunctions, httpsCallable } from "firebase/functions";



export const StoreProvider = ({ children, testStore }: React.PropsWithChildren<{ testStore?: IRootStore }>) => {

    const firebase = useFirebase();

    if (testStore) {
        return (
            <StoreContext.Provider value={testStore}>
                {children}
            </StoreContext.Provider>
        );
    }

    const auth = getAuth(firebase);
    const firestore = getFirestore(firebase);
    const functions = getFunctions(firebase);

    const store = (window as any)["store"] = testStore || new Store({
        auth,
        firestore,
        httpsCallable: (name) => httpsCallable(functions, name),
    });

    return (
        <StoreContext.Provider value={store}>
            {children}
        </StoreContext.Provider>
    );
}
