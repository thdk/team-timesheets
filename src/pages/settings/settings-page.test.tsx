import React from "react";
import path from "path";
import fs from "fs";

import { SettingsPage } from "./settings-page";
import { render } from "@testing-library/react";
import { initializeTestApp, loadFirestoreRules } from "@firebase/rules-unit-testing";
import { IRootStore, Store } from "../../stores/root-store";
import { clearFirestoreData } from "firestorable/lib/utils";
import { useStore } from "../../contexts/store-context";

import type firebase from "firebase";

jest.mock("../../contexts/store-context");

const projectId = "settings-page";
const app = initializeTestApp({
    projectId,
});

let store: IRootStore;
beforeAll(async () => {
    await loadFirestoreRules({
        projectId,
        rules: fs.readFileSync(path.resolve(__dirname, "../../../firestore.rules.test"), "utf8"),
    });
});

beforeEach(async () => {
    store = new Store({
        firestore: app.firestore(),
    });

    await store.auth.addDocument({
        uid: "user-1",
        divisionId: "",
        name: "User 1",
        recentProjects: [],
        roles: { user: true },
        tasks: new Map(),
        email: "user@timesheets.com",
    }, "user-1");

    store.auth.setUser({
        uid: "user-1",
    } as firebase.User);

    (useStore as jest.Mock<ReturnType<typeof useStore>>).mockReturnValue(store);
});

afterEach(async () => {
    store.dispose();
    await clearFirestoreData(projectId);
});

afterAll(() => app.delete());

jest.mock('../../containers/tasks/list', () => ({
    TaskList: () => <>Tasks-Content</>,
}));
jest.mock('../../containers/users/list', () => ({
    UserList: () => <>Users-Content</>,
}));
jest.mock('../../containers/clients/list', () => ({
    ClientList: () => <>Clients-Content</>,
}));
jest.mock('../../containers/teams/list', () => ({
    TeamList: () => <>Teams-Content</>,
}));

describe("SettingsPage", () => {

    it("should render", () => {
        const { asFragment } = render(
            <SettingsPage />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
