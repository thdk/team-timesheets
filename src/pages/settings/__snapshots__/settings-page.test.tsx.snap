// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsPage should render 1`] = `
<DocumentFragment>
  <nav
    class="mdc-tab-bar"
  >
    <div
      class="mdc-tab-scroller"
    >
      <div
        class="mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll"
        style="margin-bottom: 0px;"
      >
        <div
          class="mdc-tab-scroller__scroll-content"
        >
          <button
            class="mdc-tab"
          >
            <div
              class="mdc-tab__content"
            >
              <span
                class="mdc-tab__text-label"
              >
                Tasks
              </span>
            </div>
            <span
              class="mdc-tab-indicator"
            >
              <span
                aria-hidden="true"
                class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
              />
            </span>
            <div
              class="mdc-tab__ripple "
            />
          </button>
          <button
            class="mdc-tab"
          >
            <div
              class="mdc-tab__content"
            >
              <span
                class="mdc-tab__text-label"
              >
                Clients
              </span>
            </div>
            <span
              class="mdc-tab-indicator"
            >
              <span
                aria-hidden="true"
                class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
              />
            </span>
            <div
              class="mdc-tab__ripple "
            />
          </button>
          <button
            class="mdc-tab"
          >
            <div
              class="mdc-tab__content"
            >
              <span
                class="mdc-tab__text-label"
              >
                Connections
              </span>
            </div>
            <span
              class="mdc-tab-indicator"
            >
              <span
                aria-hidden="true"
                class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
              />
            </span>
            <div
              class="mdc-tab__ripple "
            />
          </button>
        </div>
      </div>
    </div>
  </nav>
  Tasks-Content
</DocumentFragment>
`;
